File: 055_MinimumNumberOfTapsToOpenToWaterAGarden.cpp
// Topic: Greedy / Interval Cover
// Minimum Number of Taps to Open to Water a Garden: convert ranges and greedily jump.
// Time Complexity: O(n)
// Space Complexity: O(n)

#include <bits/stdc++.h>
using namespace std;

int minTaps(int n, vector<int>& ranges)
{
    vector<int> maxReach(n + 1, 0);
    for (int i = 0; i <= n; i++)
    {
        int left = max(0, i - ranges[i]);
        int right = min(n, i + ranges[i]);
        maxReach[left] = max(maxReach[left], right);
    }
    int taps = 0, curEnd = 0, nextEnd = 0;
    for (int i = 0; i <= n; i++)
    {
        nextEnd = max(nextEnd, maxReach[i]);
        if (i == curEnd)
        {
            if (nextEnd <= i)
                return -1;
            taps++;
            curEnd = nextEnd;
            if (curEnd >= n)
                break;
        }
    }
    return taps;
}

int main()
{
    cout << minTaps(5, (vector<int>{3, 4, 1, 1, 0, 0})) << "\n";  // 1
    cout << minTaps(3, (vector<int>{0, 0, 0, 0})) << "\n";        // -1
    return 0;
}

File: 055_MinimumNumberOfTapsToOpenToWaterAGarden.py
"""
Topic: Greedy / Interval Cover
Problem: Minimum Number of Taps to Open to Water a Garden
Description: Transform taps to intervals and use greedy jump (like jump game II).
Time: O(n)
Space: O(n)
"""

from typing import List


def minTaps(n: int, ranges: List[int]) -> int:
    max_reach = [0] * (n + 1)
    for i, r in enumerate(ranges):
        left = max(0, i - r)
        right = min(n, i + r)
        max_reach[left] = max(max_reach[left], right)

    taps = 0
    curr_end = 0
    next_end = 0
    i = 0
    while curr_end < n:
        while i <= curr_end:
            next_end = max(next_end, max_reach[i])
            i += 1
        if next_end == curr_end:
            return -1
        taps += 1
        curr_end = next_end
    return taps


if __name__ == '__main__':
    print(minTaps(5, [3,4,1,1,0,0]))  # 1
    print(minTaps(3, [0,0,0,0]))  # -1
